### PR TITLE
Rare Kaon decay contraction code

### PR DESCRIPTION
**Pull request by Lanny91**

### CODE REVIEW

File: extras/Hadrons/Application.cc [hunk 1]
```diff
@@ -91,6 +91,11 @@ const Application::GlobalPar & Application::getPar(void)
     return par_;
 }
 
+bool Application::hasModule(const std::string name)
+{
+    return env().hasModule(name);
```
Comment (Lanny91) [inline extras/Hadrons/Application.cc]: Agreed and removed.

File: extras/Hadrons/Modules/MContraction/Meson.hpp [hunk 2]
```diff
@@ -111,23 +162,46 @@ void TMeson<FImpl1, FImpl2>::execute(void)
...
     
-    XmlWriter             writer(par().output);
-    PropagatorField1      &q1 = *env().template getObject<PropagatorField1>(par().q1);
-    PropagatorField2      &q2 = *env().template getObject<PropagatorField2>(par().q2);
-    LatticeComplex        c(env().getGrid());
-    Gamma                 gSrc(par().gammaSource), gSnk(par().gammaSink);
-    Gamma                 g5(Gamma::Algebra::Gamma5);
-    std::vector<TComplex> buf;
-    Result                result;
+    XmlWriter              writer(par().output);
```
Comment (Lanny91) [inline extras/Hadrons/Modules/MContraction/Meson.hpp]: Is this implemented yet? I don't see the definition of CorrWriter anywhere else in the current branch. I've left this for now.

File: extras/Hadrons/Modules/MContraction/Meson.hpp [hunk 3]
```diff
@@ -111,23 +162,46 @@ void TMeson<FImpl1, FImpl2>::execute(void)
...
+    p  = strToVec<Real>(par().mom);
+    LatticeComplex         ph(env().getGrid()), coor(env().getGrid());
+    Complex                i(0.0,1.0);
+    ph = zero;
+    for(unsigned int mu = 0; mu < env().getNd(); mu++)
+    {
+        LatticeCoordinate(coor, mu);
+        ph = ph + p[mu]*coor*((1./(env().getGrid()->_fdimensions[mu])));
+    }
+    ph = exp(-(Real)(2*M_PI)*i*ph);
```
Comment (Lanny91) [inline extras/Hadrons/Modules/MContraction/Meson.hpp]: I am sure that it shouldn't be there! Fixed.

File: extras/Hadrons/Modules/MContraction/Meson.hpp [hunk 4]
```diff
@@ -111,23 +162,46 @@ void TMeson<FImpl1, FImpl2>::execute(void)
...
-    for (unsigned int t = 0; t < buf.size(); ++t)
+    parseGammaString(gammaList);
+
+    result.resize(gammaList.size());
+    for (unsigned int i = 0; i < result.size(); ++i)
     {
-        result.corr[t] = TensorRemove(buf[t]);
+        Gamma gSnk(gammaList[i].first);
+        Gamma gSrc(gammaList[i].second);
+        c = trace((g5*gSnk)*q1*(gSrc*g5)*adj(q2))*ph;
```
Comment (Lanny91) [inline extras/Hadrons/Modules/MContraction/Meson.hpp]: Done.

File: extras/Hadrons/Modules/MContraction/Meson.hpp [hunk 5]
```diff
@@ -111,23 +162,46 @@ void TMeson<FImpl1, FImpl2>::execute(void)
...
+        sliceSum(c, buf, Tp);
+
+        result[i].gamma_snk = gammaList[i].first;
+        result[i].gamma_src = gammaList[i].second;
+        result[i].corr.resize(buf.size());
+        for (unsigned int t = 0; t < buf.size(); ++t)
+        {
+            result[i].corr[t] = TensorRemove(buf[t]);
+        }
+        write(writer, "meson", result[i]);
```
Comment (Lanny91) [inline extras/Hadrons/Modules/MContraction/Meson.hpp]: Fixed.

File: extras/Hadrons/Modules/MContraction/WeakHamiltonianEye.cc [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+    std::vector<PropagatorField> S_loop(ndim, tmp1);
+    std::vector<LatticeComplex>  E_body(ndim, tmp2);
+    std::vector<LatticeComplex>  E_loop(ndim, tmp2);
+
+    // Setup for S-type contractions.
+    for (int mu = 0; mu < ndim; ++mu)
+    {
+        S_body[mu] = MAKE_SE_BODY(q1, q2, q3, Gamma::gmu[mu]) -
+                     MAKE_SE_BODY(q1, q2, q3, Gamma::gmu[mu]*g5);
+        S_loop[mu] = MAKE_SE_LOOP(q4, Gamma::gmu[mu]) -
```
Comment (Lanny91) [inline extras/Hadrons/Modules/MContraction/WeakHamiltonianEye.cc]: Fixed.

File: extras/Hadrons/Modules/MContraction/WeakHamiltonianNonEye.cc [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    LatticeComplex               tmp2(env().getGrid());
+    std::vector<PropagatorField> C_i_side_loop(ndim, tmp1);
+    std::vector<PropagatorField> C_f_side_loop(ndim, tmp1);
+    std::vector<LatticeComplex>  W_i_side_loop(ndim, tmp2);
+    std::vector<LatticeComplex>  W_f_side_loop(ndim, tmp2);
+
+    // Setup for C-type contractions.
+    for (int mu = 0; mu < ndim; ++mu)
+    {
+        C_i_side_loop[mu] = MAKE_CW_SUBDIAG(q1, q2, Gamma::gmu[mu]) -
```
Comment (Lanny91) [inline extras/Hadrons/Modules/MContraction/WeakHamiltonianNonEye.cc]: Fixed.

File: extras/Hadrons/Modules/MSink/Wall.hpp [hunk 8]
```diff
@@ -0,0 +1,144 @@
+/*************************************************************************************
```
Comment (Lanny91) [inline extras/Hadrons/Modules/MSink/Wall.hpp]: Dropped.

File: lib/serialisation/BaseIO.h [hunk 9]
```diff
@@ -138,6 +138,54 @@ namespace Grid {
     unsigned int               dimInd_{0};
   };
   
+  // Pair IO utilities /////////////////////////////////////////////////////////
+  // helper function to parse input in the format "<obj1 obj2>"
+  template <typename T1, typename T2>
```
Comment (Lanny91) [inline lib/serialisation/BaseIO.h]: I thought about a regex, but this seemed to me to be a much simpler implementation, as I don't think it is possible to directly use a regex on the istream. This code does avoid the spacing issues that you mentioned, but doesn't check that you only have 2 objects between your delimiters. I could get around this by using a regex on the string I've plucked from between the `<` and `>` delimiters.

File: tests/hadrons/Test_hadrons_rarekaon.cc [hunk 10]
```diff
@@ -0,0 +1,463 @@
+/*******************************************************************************
```
Comment (Lanny91) [inline tests/hadrons/Test_hadrons_rarekaon.cc]: Removed for now.

File: extras/Hadrons/Modules/MContraction/Meson.hpp (hunk 2 above)
Comment (aportelli) [inline extras/Hadrons/Modules/MContraction/Meson.hpp]: Ah, it was accidentally committed into the `qed-fvol` feature, my bad. I have now added it to `hadrons`, you can pull and make the change.

Comment (Lanny91) [inline extras/Hadrons/Modules/MContraction/Meson.hpp]: Done!

**Comment by aportelli** [comment]
All good to me, that's great work thanks!

**Pull request merged**